Plutarch/Verify.hs

```diff
@@ -0,0 +1,99 @@
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.Verify (
+  PTryFrom (ptryFrom),
+  PTryFromRecur (ptryFromRecur),
+) where
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pdata,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  phoistAcyclic,
+  plam,
+  (#),
+  (#$),
+  type (:-->),
+ )
+import Plutarch.List (pmap)
+
+import Plutarch.Lift (PLift)
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+{- |
+    Each POpaque can be of any representation as it represents
+    the universe of Types. With PTryFrom we establish trust
+    in the Data by verifying the requested type matches its
+    representation.
+-}
+class PTryFrom (a :: PType) where
+  ptryFrom :: Term s (PData :--> a)
```

Comment:
This should be a multiparam class IMO,
and we can have `ptryFromData` as a synonym for convenience

Reply:
I will have to think about what the application of it being a MultiParamTypeClass would be (As normally you don't want to have Types that are unverified anyway and `PData` / `POpaque` are the only types that directly occur to me that this could be needed for.)

Reply:
this is also safe for `PlutusType a => Term s a -> Term s (PInner a)`. We already have a function for this (`pto`) but it might be worth including in the conversion typeclasses.

Going further, `Term s (PInner a) -> Term s a` is also safe in certain cases but this may be harder to accomodate for. The conversion itself is given by `punsafeFrom`

Reply:
> Going further, Term s (PInner a) -> Term s a is also safe in certain cases but this may be harder to accomodate for. The conversion itself is given by punsafeFrom

I couldn't see any, but if there is a rule for when  `punsafeFrom` is safe that is implementable for `PTryFrom`, that would be very handy because we would get the `PFrom` instance for free, if not there will just be a `PFrom` instance that aliases to `pto`